guide/english/certifications/javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code/index.md

```diff
@@ -3,8 +3,95 @@ title: Understand the Hazards of Using Imperative Code
+
+#### Part 4
+```
+.join(workWindow.tabClose(1).tabOpen());
+```
+This part would close second tab (index 1) in the workWindow **['GMail', 'Inbox', 'Work mail', 'Docs', 'freeCodeCamp']**, which would be 'Inbox', and after that push a 'new tab' to the array, returning **['GMail', 'Inbox', 'Work mail', 'Docs', 'freeCodeCamp', 'new tab']** and adding it to the main array for a complete array which should look like this:  
+**['FB', 'Gitter', 'Reddit', 'Twitter', 'Medium', 'new tab', 'Netflix', 'YouTube', 'Vine', 'GMail', 'Work mail', 'Docs', 'freeCodeCamp', 'new tab']**, as mentioned in the instructions.
+
+
+
+If we compare the requested array and the one we recieved after running the initial code, we can see that value 'Vine' is omitted. Therefore, we need to see what is the cause of this. We can see that value 'Vine' is a value of an array **videoWindow**.  
+Now that we know this, we will check the operations done on that array. This is done in Part 3:
+```
+.join(videoWindow.tabClose(2)) // Close third tab in video window, and join
+```
+
+We can see that **tabClose()** is performed on this array, so we will analyze the code contained in that method.
+```
+tabClose = function (index) {
+ 1. var tabsBeforeIndex = this.tabs.splice(0, index); // get the tabs before the tab
+ 2. var tabsAfterIndex = this.tabs.splice(index); // get the tabs after the tab
+
+ 3. this.tabs = tabsBeforeIndex.concat(tabsAfterIndex); // join them together 
+ 4. return this;
+ };
+ ```
+ For an index, we will take 2 - as done in the challenge.
+ 1. At the beginning, our array **videoWindow** looks like this:   
+ **['Netflix', 'YouTube', 'Vimeo', 'Vine']**.
+ 2. After first line is executed, variable tabsBeforeIndex will take 2 (index) values starting from 0 (splice(0,2))  and create  a new array containing them.  
+ Arrays now look like this:  
+ tabsBeforeIndex: **['Netflix', 'YouTube']**  
+ videoWindow: **['Vimeo', 'Vine']**   
+ 
+ You can already see why splice() can be very hazardous, as it always modifies the array it is executed upon and returns that modified array - it is not deterministic.  
+ 
+ 3. After second line is executed, tabsAfterIndex should take values starting from index (in this case 2) and remove them from videoWindow array. As we can see that in the current state (**['Vimeo', 'Vine']**), videoWindow does not have index 2 so empty array is returned. Final state:  
+ tabsBeforeIndex: **['Netflix', 'YouTube']**  
+ videoWindow: **['Vimeo', 'Vine']**  
+ tabsAfterIndex: **[]**  
+ 
+ After the third line and concatenation the returned array is the same as tabsBeforeIndex, which results in both 'Vimeo' and 'Vine' values not being in the array.  
+ 
+ ### Solution
+ 
+ In order for the method tabClose to work properly, 
+ ```
+ var tabsAfterIndex = this.tabs.splice(index);
+ ```
+ should be replaced with
+ ```
+ var tabsAfterIndex = this.tabs.splice(1);
```

Comment:
I will edit it to use slice() in that case. I was under the impression that splice should remain so it can demonstrate the possible side effect it can have.

Reply:
Ideally the thing would have more than one test, but it's also written like we don't expect an actual fix.

Reply:
Should I completely rewrite the solution or just add Solution #2 using slice?

Reply:
Ah, good idea!